raft.c

```diff
@@ -2354,6 +2369,94 @@ void applyShardGroupChange(RedisRaftCtx *rr, raft_entry_t *entry)
     }
 }
 
+void replaceShardGroups(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    // 1. reset sharding info
+    ShardingInfo * si = rr->sharding_info;
+
+    ShardGroup *local = NULL;
+    if (si->shard_group_map != NULL) {
+        RedisModuleDictIter *iter = RedisModule_DictIteratorStartC(si->shard_group_map, "^", NULL, 0);
+
+        char * key;
+        size_t key_len;
+        ShardGroup *data;
+
+        while ((key = RedisModule_DictNextC(iter, &key_len, (void **) &data)) != NULL) {
+            /* local shardgroup will not have a filled in id */
+            if (*data->id != 0) {
+                RedisModule_Free(data);
+            } else {
+                local = data;
+            }
+        }
+        RedisModule_DictIteratorStop(iter);
+        RedisModule_FreeDict(rr->ctx, si->shard_group_map);
+        si->shard_group_map = NULL;
+
+        RedisModule_Free(si->shard_group_map);
+    }
+
+    /* if we didn't identify the local shard group, we have a problem */
+    RedisModule_Assert(local != NULL);
+
+    si->shard_group_map = RedisModule_CreateDict(rr->ctx);
+    RedisModule_DictSetC(si->shard_group_map, "", 0, local);
+    si->shard_groups_num = 1;
+    for(int i = 0; i <= REDIS_RAFT_HASH_MAX_SLOT; i++) {
+        if (si->hash_slots_map[i] != local) {
+            si->hash_slots_map[i] = NULL;
+        }
+    }
+
+    /* 2. iterate over payloads
+     * payload structure
+     * "# shard groups:payload1 len:payload1:....:payload n len:payload n:"
+     */
+    char * payload = entry->data;
+    char * pPayload = strchr(payload, ':');
+    *pPayload = 0;
+
+    char * endptr;
+    size_t num_payloads = strtoul(payload, &endptr, 10);
+    // verify we read all bytes belong to number */
+    RedisModule_Assert(endptr == pPayload);
+    payload = pPayload + 1;
+
+    for (int i = 0; i < num_payloads; i++) {
+        pPayload = strchr(payload, ':');
+        *pPayload = 0;
+        size_t payload_len = strtoul(payload, &endptr, 10);
+        RedisModule_Assert(endptr == pPayload);
+        payload = pPayload + 1;
+
+        ShardGroup sg;
+        if (ShardGroupDeserialize(payload, payload_len, &sg) != RR_OK) {
+            LOG_ERROR("Failed to deserialize shardgroup payload: [%.*s]", (int) payload_len, payload);
+            return;
+        }
+
+        if (!memcmp(rr->snapshot_info.dbid, sg.id, sizeof(sg.id))) {
+            /* 2a. if payload is for local shardgroup, skip */
+        } else {
+            /* 2b. if payload is for remote shardgroup, deserialize and add with ShardingInfoAddShardGroup() */
+            RedisModule_Assert(ShardingInfoAddShardGroup(rr, &sg) == RR_OK);
```

Comment:
sg will leak here as well?

Reply:
well, its an assert it wont really leak, right?  this is a crasher.

Reply:
fix above should take care of main case, unsure this case matters.

Reply:
Yeah, seems resolved